Mathlib/NumberTheory/Ostrowski.lean

```diff
@@ -55,4 +57,213 @@ lemma equiv_on_Nat_iff_equiv : (∃ c : ℝ, 0 < c ∧ (∀ n : ℕ , (f n)^c =
+  _ ≤ f (a * p ^ k) + f (b * m ^ k) := f.add_le' (a * p ^ k) (b * m ^ k)
+  _ ≤ 1 * (f p) ^ k + 1 * (f m) ^ k := by
+    simp only [map_mul, map_pow, le_refl]
+    gcongr
+    all_goals rw [← MulRingNorm.apply_natAbs_eq]; apply bdd
+  _ = (f p) ^ k + (f m) ^ k := by simp only [one_mul]
+  _ < 1 := by
+    rw [← add_halves (a:=1)]
+    apply add_lt_add
+    · exact le_half (f p) hp0 hp1 le_sup_left
+    · apply le_half (f m) _ hm le_sup_right
+      apply map_pos_of_ne_zero
+      intro m0
+      apply hpm
+      rw_mod_cast [m0]
+      exact dvd_zero p
+
+-- ## Step 4: |p| = p ^ (- t) for some positive real t
+
+lemma exists_pos_mulRingNorm_eq_pow_neg : ∃ (t : ℝ), 0 < t ∧ f p = p ^ (-t) := by
+  use - logb p (f p)
+  have pprime : Nat.Prime p :=
+    Prime.nat_prime (is_prime_of_smallest_Nat_zero_lt_mulRingNorm_lt_one p hp0 hp1 hmin)
+  constructor
+  · rw [Left.neg_pos_iff]
+    apply logb_neg _ hp0 hp1
+    exact_mod_cast Nat.Prime.one_lt pprime
+  · rw [neg_neg]
+    apply (rpow_logb (by exact_mod_cast Nat.Prime.pos pprime) _ hp0).symm
+    simp only [ne_eq, Nat.cast_eq_one,Nat.Prime.ne_one pprime, not_false_eq_true]
+
+-- ## Non-archimedean case: end goal
+/--
+  If `f` is bounded and not trivial, then it is equivalent to a p-adic absolute value.
+-/
+theorem mulRingNorm_equiv_padic_of_bounded :
+    ∃ p, ∃ (hp : Fact (Nat.Prime p)), MulRingNorm.equiv f (mulRingNorm_padic p) := by
+  obtain ⟨p, hfp, hmin⟩ := exists_smallest_Nat_zero_lt_mulRingNorm_lt_one hf_nontriv bdd
+  use p
+  have hprime : Prime p := is_prime_of_smallest_Nat_zero_lt_mulRingNorm_lt_one p hfp.1 hfp.2 hmin
+  have hprime_fact : Fact (Nat.Prime p) := fact_iff.2 (Prime.nat_prime hprime)
+  use (hprime_fact)
+  obtain ⟨t, h⟩ := exists_pos_mulRingNorm_eq_pow_neg p hfp.1 hfp.2 hmin
+  rw [← equiv_on_Nat_iff_equiv]
+  use (t⁻¹)
+  refine ⟨by simp only [one_div, inv_pos, h.1], ?_⟩
+  have ht : t⁻¹ ≠ 0 := by
+    apply inv_ne_zero
+    linarith only [h.1]
+  intro n
+  by_cases hn : n = 0 -- Separate cases n=0 and n ≠ 0
+  · simp only [hn, CharP.cast_eq_zero, map_zero, ne_eq, ht, not_false_eq_true, zero_rpow]
+  · push_neg at hn
+    /- Any natural number can be written as a power of p times a natural number not divisible
+    by p  -/
+    rcases Nat.exists_eq_pow_mul_and_not_dvd hn p (Nat.Prime.ne_one (Prime.nat_prime hprime))
+      with ⟨e, m, hpm, hnpm⟩
+    rw [hnpm]
+    simp only [Nat.cast_mul, Nat.cast_pow, map_mul, map_pow, mulRingNorm_eq_padic_norm,
+      padicNorm.padicNorm_p_of_prime, Rat.cast_inv, Rat.cast_natCast, inv_pow,
+      mulRingNorm_eq_one_of_not_dvd bdd p hfp.1 hfp.2 hmin m hpm, h.2]
+    rw [← padicNorm.nat_eq_one_iff] at hpm
+    rw [hpm]
+    simp only [mul_one, Rat.cast_one]
+    rw [← rpow_natCast_mul (rpow_nonneg (Nat.cast_nonneg p) (-t)),
+      ← rpow_mul (Nat.cast_nonneg p), mul_comm ↑e t⁻¹, ← mul_assoc, neg_mul,
+      mul_inv_cancel (by linarith only [h.1]), neg_mul, one_mul, rpow_neg (Nat.cast_nonneg p),
+      rpow_natCast]
```

Comment:
```suggestion
  rcases eq_or_ne n 0 with rfl | hn -- Separate cases n=0 and n ≠ 0
  · simp only [Nat.cast_zero, map_zero, ne_eq, ht, not_false_eq_true, zero_rpow]
  · /- Any natural number can be written as a power of p times a natural number not divisible
    by p  -/
    rcases Nat.exists_eq_pow_mul_and_not_dvd hn p hprime.ne_one with ⟨e, m, hpm, rfl⟩
    simp only [Nat.cast_mul, Nat.cast_pow, map_mul, map_pow, mulRingNorm_eq_padic_norm,
      padicNorm.padicNorm_p_of_prime, Rat.cast_inv, Rat.cast_natCast, inv_pow,
      mulRingNorm_eq_one_of_not_dvd bdd p hfp.1 hfp.2 hmin hpm, h.2]
    rw [← padicNorm.nat_eq_one_iff] at hpm
    simp only [← rpow_natCast, p.cast_nonneg, ← rpow_mul, mul_one, ← rpow_neg, hpm, cast_one]
    congr
    field_simp [h.1.ne']
    ring
```
These computations with real (or complex) powers could use a bit more automation...
I've tried to reduce to an equality of exponents that can be dealt with using `field_simp; ring`.

Reply:
Thanks! I didn't know about `field_simp`...